subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1277,6 +1277,11 @@ int ztls_accept_ctx(struct net_context *parent, struct sockaddr *addr,
 	}
 
 	child = k_fifo_get(&parent->accept_q, K_FOREVER);
+	if (child == NULL) {
+		z_free_fd(fd);
+		errno = EAGAIN;
```

Comment:
I just wonder if this is proper errno value when the listening socket is closed. In the calling app, a new socket needs to be created in that case. Would `EBADF` be more proper, at least the `accept()` manual page says `EBADF  sockfd is not an open file descriptor.`, what do you think?

Reply:
EBADF makes sense I guess. I used the same error returned by sockets.c but this takes non-blocking sockets in mind I guess.

@rlubos By the way side question but are there plans to support async TLS sockets in the future?

Reply:
I agree, since `EAGAIN` should only be used with non-blocking sockets, it might be confusing to receive it on a blocking one, plus ambiguous with non-blocking sockets. `EBADF` sounds like a good alternative.

Reply:
> By the way side question but are there plans to support async TLS sockets in the future?

If you mean async handshake, not on my roadmap currently I'm afriad. As for regular `send`/`recv` it should work I believe (unless some bugs slipped in over time).

Reply:
EAGAIN has very specific meaning (non-blocking operation was requested, and operation can't complete without blocking), and shouldn't be used for anything else. Even looking at the 2 last lines makes it clear that EAGAIN is wrong choice here: you can't really "free" fd and return EAGAIN, the whole idea of EAGAIN that you're supposed to call the same (or another) function on that fd again, so you can't close or free it. (EAGAIN isn't even an "error" per se, it's "transient condition").

Reply:
I'll change it but just FYI sockets.c : zsock_accept_ctx uses the exact same code ;)

Reply:
> > By the way side question but are there plans to support async TLS sockets in the future?
> 
> If you mean async handshake, not on my roadmap currently I'm afriad. As for regular `send`/`recv` it should work I believe (unless some bugs slipped in over time).

I meant more of having accept be non-blocking, currently it blocks so you cannot really have a single loop do the accept / send / recv thing in parallel if I'm not mistaken. Does what I say makes sense?

Reply:
> If you mean async handshake, not on my roadmap currently I'm afriad.

To properly implement async TLS handshake, you really need to do it in "userspace", instead of layering-violation implementation you guys have now, contrary to warnings of yours truly ;-). The problem, as you imagine, lies in the fact of not enough selectivity in non-blocking return codes in POSIX stream (fd) model. To support generalized streams like TLS, you should be prepared to a situation when write() requires blocking for read operations, and vice versa. POSIX offers just EAGAIN to convey that write() needs blocking for write(), or read() - blocking for read(), no return code for the "cross" situation.

(And I can very well imagine that you'll persist in your wrongdoing and will introduce EAGAIN_RD and EAGAIN_WR, which will only further violate POSIX API contract.)

Reply:
@pfalcon wrote:

> Even looking at the 2 last lines makes it clear that EAGAIN is wrong choice here: you can't really "free" fd and return EAGAIN,

Ok, let's call this ~~brain f%rt~~ uncareful looking: here you free the newly allocated fd, so the above tirade doesn't apply, sorry for the noise.

Reply:
@lbise:

> I'll change it but just FYI sockets.c : zsock_accept_ctx uses the exact same code ;)

Ok, gotcha. @lbise, can I please ask you to investigate this issue to get to the bottom of it - code up a Linux sample, check what it does, and do the same for both cases (and add code comment to not return to it in the future)? Thanks in advance.

Reply:
> I meant more of having accept be non-blocking, currently it blocks so you cannot really have a single loop do the accept / send / recv thing in parallel if I'm not mistaken. Does what I say makes sense?

@lbise You're right, the `accept` call seems to ignore `O_NONBLOCK` option, and always blocks waiting for an incoming connection (this should be an easy fix though). As an alternative, you could `poll` the listening socket, and call `accept` only when a new connection is present (i. e. POLLIN is returned).

Reply:
> You're right, the accept call seems to ignore O_NONBLOCK option, and always blocks waiting for an incoming connection (this should be an easy fix though).

So, is this defect of TLS implementation specifically? Because otherwise we have https://github.com/zephyrproject-rtos/zephyr/issues/6446 closed.

Reply:
Correct, it was fixed for regular sockets in https://github.com/zephyrproject-rtos/zephyr/commit/0b1034aee20b15093c38399608186386d8ddd624

Reply:
> Correct, it was fixed for regular sockets in [0b1034a](https://github.com/zephyrproject-rtos/zephyr/commit/0b1034aee20b15093c38399608186386d8ddd624)

Hmm, in that commit, I was actually returning EAGAIN which is probably not ok if the socket is blocking one. We need to fix that one.